src/vms/pvm/etna-builder/spend.ts

```diff
@@ -0,0 +1,473 @@
+        // We couldn't spend this UTXO, so we skip to the next one.
+        continue;
+      }
+
+      spendHelper.addInput(
+        utxo,
+        // TODO: Verify this.
+        // TransferableInput.fromUtxoAndSigindicies(utxo, inputSigIndices),
+        new TransferableInput(
+          utxo.utxoId,
+          utxo.assetId,
+          new TransferInput(out.amt, Input.fromNative(inputSigIndices)),
+        ),
+      );
+
+      const excess = spendHelper.consumeAsset(assetId, out.amount());
+
+      if (excess === 0n) {
+        continue;
+      }
+
+      // This input had extra value, so some of it must be returned as change.
+      spendHelper.addChangeOutput(
+        // TODO: Verify this.
+        new TransferableOutput(
+          utxo.assetId,
+          new TransferOutput(
+            new BigIntPr(excess),
+            OutputOwners.fromNative(spendOptions.changeAddresses),
+          ),
+        ),
+      );
+    }
+
+    for (const utxo of utxosByAVAXAssetId.requested) {
+      const requiredFee = spendHelper.calculateFee();
+
+      // If we don't need to burn or stake additional AVAX and we have
+      // consumed enough AVAX to pay the required fee, we should stop
+      // consuming UTXOs.
+      if (
+        !spendHelper.shouldConsumeAsset(context.avaxAssetID) &&
+        excessAVAX >= requiredFee
+      ) {
+        break;
+      }
+
+      const [error, out] = unwrapOutput(utxo.output);
+
+      if (error) {
+        return [error, null];
+      }
+
+      const { sigIndicies: inputSigIndices } =
+        matchOwners(
+          out.outputOwners,
+          [...fromAddresses],
+          spendOptions.minIssuanceTime,
+        ) || {};
+
+      if (inputSigIndices === undefined) {
+        // We couldn't spend this UTXO, so we skip to the next one.
+        continue;
+      }
+
+      spendHelper.addInput(
+        utxo,
+        // TODO: Verify this.
+        new TransferableInput(
+          utxo.utxoId,
+          utxo.assetId,
+          new TransferInput(out.amt, Input.fromNative(inputSigIndices)),
+        ),
+      );
+
+      const excess = spendHelper.consumeAsset(
+        context.avaxAssetID,
+        out.amount(),
+      );
+
+      excessAVAX += excess;
+
+      // If we need to consume additional AVAX, we should be returning the
+      // change to the change address.
+      ownerOverride = OutputOwners.fromNative(spendOptions.changeAddresses);
+    }
+
+    // Verify
+    const verifyError = spendHelper.verifyAssetsConsumed();
+    if (verifyError) {
+      return [verifyError, null];
+    }
+
+    const requiredFee = spendHelper.calculateFee();
+
+    if (excessAVAX < requiredFee) {
+      throw new Error(
+        `Insufficient funds: provided UTXOs need ${
+          requiredFee - excessAVAX
+        } more nAVAX (${context.avaxAssetID})`,
+      );
+    }
+
+    // NOTE: This logic differs a bit from avalanche go because our classes are immutable.
+    spendHelper.addOutputComplexity(
```

Comment:
What is this output? Can you explain why we are adding it here?

Reply:
This note is just referring to the structure of the code, not the end result (logic).

It was mostly a note for myself just so I didn't get confused looking back at it.

It won't be relevant once this function gets refactored to be more idiomatic.